colossalai/inference/batch_bucket.py

```diff
@@ -0,0 +1,459 @@
+        fd_interm_tensor=None,
+        device=None,
+        dtype=torch.float16,
+    ):
+        self.num_heads = num_heads
+        self.head_dim = head_dim
+        self.max_batch_size = max_batch_size
+        self.max_length = max_length  # in + out len
+        self.block_size = block_size
+        self.kv_max_split_num = kv_max_split_num  # Hint used for flash decoding
+        self.fd_interm_tensor = fd_interm_tensor
+        self.device = device or get_current_device()
+        self.dtype = dtype
+
+        self._current_batch_size = 0
+        self._sequences_dict = dict()
+        self._sequences_indexes = dict()  # deque(maxlen=self.max_batch_size)
+        self._sequence_lengths = torch.zeros((self.max_batch_size,), dtype=torch.int32)
+        self._sequence_lengths_helper = torch.zeros_like(self._sequence_lengths)
+        max_blocks_per_seq = (self.max_length + block_size - 1) // block_size
+        self._block_tables = torch.full((self.max_batch_size, max_blocks_per_seq), -1, dtype=torch.int32)
+        self._block_tables_helper = torch.full_like(self._block_tables, -1)
+
+    @property
+    def is_empty(self):
+        return self._current_batch_size == 0
+
+    @property
+    def current_batch_size(self):
+        return self._current_batch_size
+
+    @property
+    def available_batch_size(self):
+        return self.max_batch_size - self._current_batch_size
+
+    @property
+    def block_tables(self):
+        return self._block_tables
+
+    @property
+    def seq_lengths(self):
+        return self._sequence_lengths
+
+    @property
+    def seqs_ids(self):
+        return list(self._sequences_dict.keys())
+
+    @property
+    def seqs_li(self):
+        return list(self._sequences_dict.values())
+
+    @property
+    def is_compact(self):
+        assert len(self._sequences_dict) == len(self._sequences_indexes), "BatchBucket indexing is not consistent"
+        return (
+            len(self._sequences_dict)
+            == torch.nonzero(self._sequence_lengths).view(-1).numel()
+            == torch.nonzero(self._block_tables[:, 0] >= 0).numel()
+        )
+
+    def _make_compact(self) -> None:
+        # Clean and Compress the batch based on its sequences dict.
+        # Namely,compress sequences to the front and clean the seq lengths and block tables tensors.
+        # NOTE Prevent calling this method multiple times in a single step
+        if self.is_compact:
+            return
+        valid_seq_ids = self._sequences_dict.keys()
+        valid_num = len(valid_seq_ids)
+        valid_indexes = [self._sequences_indexes[seq_id] for seq_id in valid_seq_ids]
+        assert valid_num == len(self._sequences_indexes), "BatchBucket indexing is not consistent"
+        self._sequence_lengths_helper[:valid_num] = self._sequence_lengths[valid_indexes]
+        self._sequence_lengths[:] = self._sequence_lengths_helper[:]
+        self._block_tables_helper[:valid_num, :] = self.block_tables[valid_indexes]
```

Comment:
Does the movement of the blocktable require the physical memory to move along with it?

Reply:
We modify the same physical memory here, instead of creating new ones.

Reply:
In summary, we'd like an aggregated block tables (a 2D tensor) to be used along the inference